### ISSUE TITLE
Messages marked as "failed to send" are delivered

### ISSUE THREAD

**Issue by righy** [signalapp/Signal-Android#8625]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Today I lost communication with one of my contact (contact A). I sent a message to A. The message was added to the chat, along with an exclamation mark icon and "Send failed, tap for unsecured fallback" text. I tapped the message and got a dialog asking "Fallback to unencrypted SMS? This message will not be encrypted because the recipient is no longer a Signal user. Send unsecured message? Yes/No". I tapped "no". The text changed to "Send failed, tap for details".

I tried to send a message to another Signal contact (contact B) and it got sent to the server (one tick). It looks like B uninstalled Signal because it's been a long time all my messages are not delivered to B, so I cannot confirm it fully works.

I tried to delete the "failed to send" message to A and send it again two times and got the same error each time. I contacted A through another channel and it turns out A got all my messages that appeared as "failed to send" to me. What's more, A tried to send me a last message yesterday that failed to send (without the "Fallback to unencrypted SMS") and I didn't receive it. A cannot send me any messages now because I appear as if I'm not registered on Signal ("enable signal for SMS" button instead of text field).

I tried to send myself a message from my phone but I disappeared from the list of contacts on Signal. Contact A is still on the list though.

I tried to send myself a message from Signal desktop, and it failed to send. I tried to send a message to B from Signal desktop, and it shows as sent to the server, but it doesn't show up on my phone.

### Steps to reproduce
- Try to send a message to my contact A

**Actual result:** The message appears as failed to sent even though it is actually sent, and A cannot reply to me because I appear as not on Signal.
**Expected result:** The message is sent and A can reply to me.

### Device info

**Device:** Samsung GT-N7100
**Android version:** 7.1.2
**Signal version:** 4.33.5

### Link to debug log
[Debug log from my phone](https://debuglogs.org/871180570f04fe49f286e3c29a8f8beec3243782c6245693c19d6de89c29ed01).
[Debug log from the phone of A](https://debuglogs.org/2d2958df77f5c72555eabf0a0fe34ce20f215cbf1db0c37e4d147135d44ad1d0).

**Comment by greyson-signal** [comment]
Alright, so just writing down what we know:

- Phone A is running a pre-FCM version of Signal, so this likely isn't related to the FCM migration.
- Your phone is a non-Google device, so no GCM/FCM problems there
- For whatever reason, your persistent logs don't exist, so I don't have many logs from your device, and I can't see any failed message sends
- I'm concerned that message that were described as "failed to send" made their way to the recipient
- Somehow you became unregistered, and I'm not sure how. Did you remove Google Play Services from your phone after installing Signal?

**Comment by righy** [comment]
> For whatever reason, your persistent logs don't exist, so I don't have many logs from your device, and I can't see any failed message sends

I tried to send another message to A (still failed) and [submitted a new debug log](https://debuglogs.org/301a3ff6cfb62ac638eb02f5000787792836ec925eecf1a40e530d9388bcd3ca). It looks like this log is OK, but it does not contain what I mentioned in the issue description because all that happened before the beginning of this log.

> Did you remove Google Play Services from your phone after installing Signal?

I never installed any of the Google apps.

Not sure if it's related, but I am using microg. Signal used to be registered to GCM through microg. I worked fine but I didn't like that, so 3 days ago I disabled Signal registration in microg and turned off GCM in Signal by changing `pref_gcm_disabled` to `false` in its XML config file. After restarting Signal, it asked me to disable battery optimization and ran fine for the past 3 days.

**Comment by righy** [comment]
> I'm concerned that message that were described as "failed to send" made their way to the recipient

Me too. I confirm my messages are still being received by A. Should I re-register my phone number?

**Comment by greyson-signal** [comment]
> Not sure if it's related, but I am using microg. Signal used to be registered to GCM through microg. I worked fine but I didn't like that, so 3 days ago I disabled Signal registration in microg and turned off GCM in Signal by changing `pref_gcm_disabled` to `false` in its XML config file. After restarting Signal, it asked me to disable battery optimization and ran fine for the past 3 days.

That's very much related :) If you were using GCM before and then turned it off, the server is going to take that as a hint that you've unregistered. 

Given that you had to manually mutate system-level packages and then change hidden Signal preferences, it should be clear that this isn't a supported user path :) 

> Should I re-register my phone number?

Yes.

> I tried to send another message to A (still failed) and [submitted a new debug log](https://debuglogs.org/301a3ff6cfb62ac638eb02f5000787792836ec925eecf1a40e530d9388bcd3ca).

Here's the relevant part of that log:

```java
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: [547487e3-b6db-473f-a364-313fec2d25e9] Failure (Time since submission: 2202 ms, Run attempt: 0, isStopped: false)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: org.whispersystems.signalservice.api.push.exceptions.UnregisteredUserException: org.whispersystems.signalservice.api.push.exceptions.NotFoundException: Not found
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.sendMessage(PushServiceSocket.java:227)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:851)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:222)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushTextSendJob.deliver(PushTextSendJob.java:169)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushTextSendJob.onPushSend(PushTextSendJob.java:92)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushSendJob.onSend(PushSendJob.java:86)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.SendJob.onRun(SendJob.java:56)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobmanager.Job.doWorkInternal(Job.java:111)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobmanager.Job.doWork(Job.java:64)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at androidx.work.Worker$1.run(Worker.java:84)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at java.lang.Thread.run(Thread.java:761)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: Caused by: org.whispersystems.signalservice.api.push.exceptions.NotFoundException: Not found
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.makeServiceRequest(PushServiceSocket.java:799)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.sendMessage(PushServiceSocket.java:222)
2019-02-19 14:47:03.839 GMT+08:00 W PushTextSendJob: 	... 12 more
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: [547487e3-b6db-473f-a364-313fec2d25e9] Failure (Time since submission: 2204 ms, Run attempt: 0, isStopped: false)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: org.thoughtcrime.securesms.transport.InsecureFallbackApprovalException: org.whispersystems.signalservice.api.push.exceptions.UnregisteredUserException: org.whispersystems.signalservice.api.push.exceptions.NotFoundException: Not found
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushTextSendJob.deliver(PushTextSendJob.java:172)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushTextSendJob.onPushSend(PushTextSendJob.java:92)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushSendJob.onSend(PushSendJob.java:86)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.SendJob.onRun(SendJob.java:56)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobmanager.Job.doWorkInternal(Job.java:111)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobmanager.Job.doWork(Job.java:64)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at androidx.work.Worker$1.run(Worker.java:84)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at java.lang.Thread.run(Thread.java:761)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: Caused by: org.whispersystems.signalservice.api.push.exceptions.UnregisteredUserException: org.whispersystems.signalservice.api.push.exceptions.NotFoundException: Not found
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.sendMessage(PushServiceSocket.java:227)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:851)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.api.SignalServiceMessageSender.sendMessage(SignalServiceMessageSender.java:222)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.thoughtcrime.securesms.jobs.PushTextSendJob.deliver(PushTextSendJob.java:169)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	... 9 more
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: Caused by: org.whispersystems.signalservice.api.push.exceptions.NotFoundException: Not found
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.makeServiceRequest(PushServiceSocket.java:799)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	at org.whispersystems.signalservice.internal.push.PushServiceSocket.sendMessage(PushServiceSocket.java:222)
2019-02-19 14:47:03.890 GMT+08:00 W PushTextSendJob: 	... 12 more
```

And to be clear, you're saying that you're sending messages that are marked as failed on your client, but are still received by A? If that's true, I wonder if the issue is that *you're* the one who isn't registered.

**Comment by righy** [comment]
> > Should I re-register my phone number?
> 
> Yes.

I unregistered and registered my phone number by going to advanced options, but it didn't solve the problem.

> And to be clear, you're saying that you're sending messages that are marked as failed on your client, but are still received by A?

Yes, the messages are marked as failed to send but still received by A. A sent me screenshots so I am sure about that.

**Comment by righy** [comment]
I tried to send a message to another contact (contact C). It failed just like A, but C received the message and he replied to me by SMS.

**Comment by righy** [comment]
I tried to send a message to a group chat, and the message is not being sent ("spinning dots" icon), but I see no "failed to send" error of any kind. The "spinning dots" have been spinning for hours.

**Comment by greyson-signal** [comment]
@righy  Can you verify that the messages sent to A were Signal messages and not SMS messages? SMS messages are marked with a little "open lock" icon in the corner. We went through the server code and don't see how we could get back a 404 for a message send but still send it.

Can you also confirm whether or not A is also on a rooted device that might use microg or something similar?

**Comment by righy** [comment]
> Can you verify that the messages sent to A were Signal messages and not SMS messages?

I confirm that the messages received are not SMS (no open lock icon). Just to be sure, I tried to send a message to A with airplane mode enabled and wifi on, and A received it.

> Can you also confirm whether or not A is also on a rooted device that might use microg or something similar?

I confirm that A has a first-hand phone without modifications.

Note that the problem doesn't seem to be specific to A, because it happened with C too.

**Comment by righy** [comment]
I tried to reproduce the problem with a secondary phone of mine (phone H, running Android 4.4.4) on which I created a new Signal account with a new number. As expected, phone H sees me as not on Signal. However, messages sent to phone H are delivered without the "failed to send" error and are marked with "one tick" (behaviour similar to contact B).

On phone H, I tried to install Signal without microG, with microG but disallowing Signal GCM registration, and with microG and allowing Signal GCM registration. I also tried with and without letter sealing enabled, and with Signal 4.33.5 and 4.32.8 (A's version). None of these variations changed anything.

**Comment by righy** [comment]
I was able to reproduce the problem eventually after disabling typing notifications on H (or maybe it's just because I waited some more time?).

[Debug log of my phone](https://debuglogs.org/8fadc6bd5460df5e63f94eb13337e8c2e95eb420b7bec8ece3e5799b1f9ea5af).
[Debug log of H](https://debuglogs.org/d34c1b8cbf94cc77eb4073219090b7b3ce9e10e7edee341308ff4a311f22658a) (timestamps are shifted by 7 hours compared to my phone).

What I did (timestamps are relative to my phone):
> 15:49:35 -> sending message to H, OK one tick
> 16:04:22 -> sending message to D, failed to send (D is yet another contact)
> 16:44:45 -> disabled typing notifications on phone H
> 16:45:02 -> sending message to H, failed to send
> 16:45:30 -> enabled typing notifications on phone H
> 16:45:45 -> sending message to H, failed to send
> 16:59:49 -> sending message to B, OK one tick

**Comment by righy** [comment]
I can see a pattern in the logs:
* `UnidentifiedAccessUtil: Their access key present? false | Our access key present? true | Our certificate present? true` -> message sent with one tick and received
* `UnidentifiedAccessUtil: Their access key present? true | Our access key present? true | Our certificate present? true` -> message "failed to send" but received

**Comment by righy** [comment]
If I understand correctly, the "failed to send" error shows when the message is sent with letter sealing (true/true/true in my previous message).

I tried to disable the "allow from everyone" letter sealing option on my phone, and sent a message to H. It still showed as failed but was delivered, and I could still see true/true/true messages in the log, so I restarted Signal. And then, problem completely solved: I am seen as registered on Signal and I can exchange messages again normally with A and H.

Note sure if it's related, but according to the logs, the option `pref_gcm_disabled` has been switched back to `false` after I [re-registered](#issuecomment-465296101).

**Comment by righy** [comment]
Here the [debug log from my phone](https://debuglogs.org/fb6058c73211e99dae531fc2ca90019f73d45ffe5b3f1fc44bdbefbc4ac32e1b) just after I was able to exchange messages again with A and H. Note that the server is still replying with 404 errors, and the client falls back to a different sending mechanism.

**Comment by righy** [comment]
I tried to re-enable "allow from everyone" and restarted Signal. I worked for about ten minutes, and then the problem came back ("failed to send" with delivery of message). I disabled "allow from everyone", still not working, then I re-registered, and it was working again. 

When I [first re-registered](#issuecomment-465296101), Signal was still disabled in microg, so this is probably the reason why it didn't solve the problem. This time it was enabled and re-registering solved the problem right away.

Note that having my phone and phone H next to each other while sending a "failed to send but delivered" to H allowed me to see that the message is received by H *before* the "failed to send" error is displayed on my phone. Maybe Signal sends the message in some way but thinks that it failed whereas it didn't, and then it tries an alternative way to send it which fails and it shows the error?
